src/Orchestration/Adapter/DockerCLI.php

```diff
@@ -36,26 +149,79 @@ public function run(string $image, string $name, string $entrypoint = '', string
             " --name={$name}".
             " --label {$this->namespace}-type=runtime".
             " --label {$this->namespace}-created={$time}".
-            " --volume /xxx/xxx:/tmp:rw".
+            " --volume {$mountFolder}:/tmp:rw".
+            $labelString .
             " --workdir {$workdir}".
             " ".\implode(" ", $vars).
             " {$image}".
-            " {$command}"
+            " ".implode(" ", $command)
             , '', $stdout, $stderr, 30);
 
-        return ($code === 0);
+        if (!empty($stderr) || $result !== 0) {
+            throw new DockerCLIException("Docker Error: {$stderr}");
+        }
+
+        return rtrim($stdout);
     }
 
-    public function execute(string $name, string $command, array $vars = []): bool
+    /**
+     * Execute Container
+     *
+     * @param string $name
+     * @param string[] $command
+     * @param string &$stdout
+     * @param string &$stderr
+     * @param array<string, string> $vars
+     * @param int $timeout
+     * @return bool
+     */
+    public function execute(string $name, array $command, string &$stdout = '', string &$stderr = '', array $vars = [], int $timeout = -1): bool
     {
-        $code = Console::execute("docker exec ".\implode(" ", $vars)." {$name} {$command}"
-            , '', $stdout, $stderr, 30);
+        foreach ($command as &$value) {
+            if (str_contains($value, " ")) {
+                $value = "'".$value."'";
+            }
+        }
 
-        return ($code === 0);
+        foreach ($vars as $key => &$value) {
+            $key = $this->filterEnvKey($key);
+
+            $value = \escapeshellarg((empty($value)) ? '' : $value);
+            $value = "--env {$key}={$value}";
+        }
+
+        $result = Console::execute("docker exec ".\implode(" ", $vars)." {$name} ".implode(" ", $command)
+            , '', $stdout, $stderr, $timeout);
+            
+        if ($result !== 0) {
+            if ($result == 1) {
+                throw new TimeoutException("Command timed out");
+            } else {
+                throw new DockerCLIException("Docker Error: {$stderr}");
+            }
+        }
+
+        return !$result;
     }
 
-    public function remove($name): bool
+    /**
+     * Remove Container
+     *
+     * @param string $name
+     * @param bool $force
+     * @return bool
+     */
+    public function remove(string $name, bool $force = false): bool
     {
-        return (Console::execute("docker rm {$name}", '', '', '') === 0);
+        $stdout = '';
+        $stderr = '';
+
+        $result = Console::execute("docker rm " . ($force ? '--force': '') . " {$name}", '', $stdout, $stderr);
+
+        if (!str_contains($stdout, $name)) {
+            throw new DockerCLIException("Docker Error: {$stderr}");
```

Comment:
Not sure if throwing an exception is needed, since the method will return false on failure.

Reply:
Probably better to rely on the exception so we can pass the error data, and report it to the end-developer.